sklearn/decomposition/cdnmf_fast.pyx

```diff
@@ -7,35 +7,59 @@
 
 cimport cython
 from libc.math cimport fabs
+from cython.parallel import prange
+
+
+cdef inline double _update_cdnmf_samples(unsigned n_components,double xht, double [] HHt,double[] W,double hess,unsigned t) nogil:
+    cdef double violation = 0
+    cdef double grad = 0 - xht
+    cdef double pg
+    for r in range(n_components):
+    # for(int r =0;r<n_components;r++)
+        grad += HHt[r] * W[r]
+    pg = grad
+    if (W[t] == 0) and (pg > 0):
+        pg = 0
+    violation = fabs(pg)
+    if hess != 0:
+        W[t] -= grad/hess
+        if W[t] < 0 :
+            W[t] = 0
+    return violation
 
 
 def _update_cdnmf_fast(double[:, ::1] W, double[:, :] HHt, double[:, :] XHt,
                        Py_ssize_t[::1] permutation):
     cdef double violation = 0
     cdef Py_ssize_t n_components = W.shape[1]
     cdef Py_ssize_t n_samples = W.shape[0]  # n_features for H update
-    cdef double grad, pg, hess
+    cdef double pg, hess
     cdef Py_ssize_t i, r, s, t
 
+    # print "in nogil!!"
     with nogil:
         for s in range(n_components):
             t = permutation[s]
+            # Hessian
+            hess = HHt[t, t]
```

Comment:
no need to define `hess` here. It can be done directly in `_update_cdnmf_samples`